tensorflow_probability/python/positive_semidefinite_kernels/dot_product.py

```diff
@@ -0,0 +1,127 @@
+# Copyright 2018 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Linear kernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow_probability.python.internal import dtype_util
+from tensorflow_probability.python.positive_semidefinite_kernels import positive_semidefinite_kernel as psd_kernel
+from tensorflow_probability.python.positive_semidefinite_kernels.internal import util
+
+__all__ = [
+  'Linear',
+]
+
+
+def _validate_arg_if_not_none(arg, assertion, validate_args):
+  if arg is None:
+    return arg
+  with tf.control_dependencies([assertion(arg)] if validate_args else []):
+    result = tf.identity(arg)
+  return result
+
+
+class Linear(psd_kernel.PositiveSemidefiniteKernel):
+  """Linear Kernel.
+
+    Is based on the dot product covariance function and can
+    be obtained from linear regression. This kernel, when
+    parameterizing a Gaussian Process, results in random functions that are
+    following a linear function.
+    In practice this kernel is mostly used in combination with the
+    exponentiated quadratic kernel to make it non-stationary and local.
+
+    ```none
+    k(x, y) = amplitude**2 * ( bias_variance**2 + x \cdot x')
```

Comment:
I guess more generally the kernel could be of the form

(bias_variance**2 + slope_variance** 2 * (x - c) \cdot (x - c)'
(c.f.: https://www.cs.toronto.edu/~duvenaud/cookbook/).

Any interest in adding that to this commit (I think what is here is fine in that it would be compatible with adding offset and slope_variance, but it might be nice to do this all at once)?

Reply:
Hi @srvasude the `c` parameter you suggested from the cookbook encodes the x-axis intercept. Which in my experience if you combine the linear kernel with an exponentiated quadratic is not useful since this will be captured by the locality of that kernel. An additional complexity is that the `c` parameter should be a vector in the shape of `x's` features.

However what I do think could be a valuable generalization is making the kernel polynomial by adding an optional power variable as `bias_variance**2 + slope_variance**2 * ((x) \cdot (x )')**p`.

Reply:
@srvasude I added the intercept `c` parameter in a new PR: https://github.com/tensorflow/probability/pull/229